### PR TITLE
Initial Jib detection with skaffold init

### PR DESCRIPTION
**Pull request by TadCordle**
This PR adds Jib detection to `skaffold init`. Skaffold walks the directory searching for dockerfiles, `build.gradle`s, and `pom.xml`s. If a `build.gradle` or a `pom.xml` is found, skaffold will attempt to run an internal Jib gradle task/maven goal to determine if Jib is configured on any of the projects (Jib v1.3.0 or higher is required for this to work).

If Jib is detected and `jib.to.image` is configured with an image found in the k8s manifest, the Jib builder is added to the skaffold config automatically.

If Jib is detected but `jib.to.image` is not configured to build a k8s image, the Jib project is added to the CLI prompt so the user can decide which image Jib should build, and the user's choice is added using the `-Djib.to.image=...` flag.

Issues left to solve:
- Configuring to build with Jib via CLI artifacts
- Gradle/Maven edge cases (config files that don't use the default names, unorthodox project structure, etc.)
- Generating Jib config automatically if it's missing?

Towards https://github.com/GoogleContainerTools/skaffold/issues/1264 (probably not completely fixed until there's some idea how to automatically apply Jib to maven/gradle projects).

### CODE REVIEW

File: pkg/skaffold/docker/docker_init.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	// Target image is not configured in dockerfiles
+	return ""
+}
+
+// GetPath returns the path to the dockerfile
+func (d Dockerfile) GetPath() string {
+	return string(d)
+}
+
+// ValidateDockerfile makes sure the given Dockerfile is existing and valid.
```
Comment (TadCordle) [inline pkg/skaffold/docker/docker_init.go]: Note: this function was moved from `docker/validate.go`

File: pkg/skaffold/docker/docker_init_test.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+import (
+	"path/filepath"
+	"testing"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/testutil"
+)
+
+func TestValidateDockerfile(t *testing.T) {
```
Comment (TadCordle) [inline pkg/skaffold/docker/docker_init_test.go]: Note: This was moved from `docker/validate_test.go`

File: pkg/skaffold/initializer/init.go [hunk 3]
```diff
@@ -54,6 +54,18 @@ type Initializer interface {
 	GetImages() []string
 }
 
+// InitBuilder represents a builder that can be chosen by skaffold init.
+type InitBuilder interface {
+	// GetPrompt returns the initBuilder's string representation, used when prompting the user to choose a builder.
+	GetPrompt() string
+	// GetArtifact returns the Artifact used to generate the Build Config.
```
Comment (briandealwis) [inline pkg/skaffold/initializer/init.go]: `GetArtifact creates an Artifact to be included in the generated Build Config` ?

File: pkg/skaffold/initializer/init.go [hunk 4]
```diff
@@ -54,6 +54,18 @@ type Initializer interface {
 	GetImages() []string
 }
 
+// InitBuilder represents a builder that can be chosen by skaffold init.
+type InitBuilder interface {
+	// GetPrompt returns the initBuilder's string representation, used when prompting the user to choose a builder.
+	GetPrompt() string
```
Comment (briandealwis) [inline pkg/skaffold/initializer/init.go]: `Describe`?

File: pkg/skaffold/initializer/init.go [hunk 5]
```diff
@@ -54,6 +54,18 @@ type Initializer interface {
...
+	// GetArtifact returns the Artifact used to generate the Build Config.
+	GetArtifact(image string) *latest.Artifact
+	// GetConfiguredImage returns the target image configured by the builder
+	GetConfiguredImage() string
```
Comment (briandealwis) [inline pkg/skaffold/initializer/init.go]: Disconnect between `ConfiguredImage` and `target image`?  is it ok to have an empty string, and what does it mean?

File: pkg/skaffold/initializer/init.go [hunk 6]
```diff
@@ -54,6 +54,18 @@ type Initializer interface {
 	GetImages() []string
 }
 
+// InitBuilder represents a builder that can be chosen by skaffold init.
+type InitBuilder interface {
```
Comment (briandealwis) [inline pkg/skaffold/initializer/init.go]: _Effective Go_ seems to recommend [avoiding prefixing getters with _Get_](https://golang.org/doc/effective_go.html#Getters).  I'll put some suggestions below.

**Comment by TadCordle** [comment]
I'm going to split this PR up into 2 since it got so large. Will follow-up soon.

**Pull request closed without merging**